### PR TITLE
NEWRELIC-4191 Properly link tokens so transaction is available to user code

### PR DESCRIPTION
**Pull request by jtduffy**
Insure transactions are visible to user code.

### CODE REVIEW

File: instrumentation/grpc-1.40.0/src/main/java/com/nr/agent/instrumentation/grpc/GrpcUtil.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Expire the supplied {@link Token} instance if it's non-null.
+     *
+     * @param token The {@link Token} to expire
+     */
+    public static void expireToken(Token token) {
+        if (token != null) {
+            token.expire();
+            token = null;
```
Comment (meiao) [inline instrumentation/grpc-1.40.0/src/main/java/com/nr/agent/instrumentation/grpc/GrpcUtil.java:60]: I don't think this method is very useful.
The token that gets nulled out here is the reference that is used in this method.
If while debugging, you step out of this method, you'll see that the reference that was passed to this method is still referencing the original.

File: instrumentation/grpc-1.40.0/src/main/java/io/grpc/internal/ServerStream_Instrumentation.java [hunk 2]
```diff
@@ -28,56 +29,24 @@ public abstract class ServerStream_Instrumentation {
...
+        GrpcUtil.finalizeTransaction(token, status, new Metadata());
+        GrpcUtil.setServerStreamResponseStatus(status);
 
         Weaver.callOriginal();
 
-        if (token != null) {
-            token.expire();
-            token = null;
-        }
+        GrpcUtil.expireToken(token);
```
Comment (meiao) [inline instrumentation/grpc-1.40.0/src/main/java/io/grpc/internal/ServerStream_Instrumentation.java:49]: The token still needs to be nulled out here.

**Review by meiao — CHANGES_REQUESTED** [review]

File: instrumentation/grpc-1.40.0/src/main/java/io/grpc/internal/ServerStream_Instrumentation.java [hunk 3]
```diff
@@ -28,56 +29,24 @@ public abstract class ServerStream_Instrumentation {
...
+        GrpcUtil.finalizeTransaction(token, status, metadata);
+        GrpcUtil.setServerStreamResponseStatus(status);
 
         Weaver.callOriginal();
 
-        if (token != null) {
-            token.expire();
-            token = null;
-        }
+        GrpcUtil.expireToken(token);
```
Comment (meiao) [inline instrumentation/grpc-1.40.0/src/main/java/io/grpc/internal/ServerStream_Instrumentation.java:38]: Still need to expire the token.

File: instrumentation/grpc-1.40.0/src/main/java/com/nr/agent/instrumentation/grpc/GrpcUtil.java (hunk 1 above)
Comment (jtduffy) [reply instrumentation/grpc-1.40.0/src/main/java/com/nr/agent/instrumentation/grpc/GrpcUtil.java:60]: Good catch ... I'll refactor and remove this method

**Review by meiao — APPROVED** [review]

**Review by meiao — APPROVED** [review]

**Pull request merged**